sdks/go/pkg/beam/core/runtime/harness/statemgr_test.go

```diff
@@ -0,0 +1,248 @@
+		}, {
+			name: "readResponseChannelDeleted",
+			caseFn: func(t *testing.T, c *StateChannel, client *fakeStateClient) error {
+				go func() {
+					req := <-client.send // Send should succeed.
+
+					c.mu.Lock()
+					ch := c.responses[req.Id]
+					delete(c.responses, req.Id)
+					c.mu.Unlock()
+
+					resp := &pb.StateResponse{
+						Id: req.Id,
+					}
+					client.recv <- resp
+					// unblock Send.
+					ch <- resp
+				}()
+				_, err := c.Send(&pb.StateRequest{})
+				return err
+			},
+		}, {
+			name: "writeEOF",
+			caseFn: func(t *testing.T, c *StateChannel, client *fakeStateClient) error {
+				go func() {
+					client.sendErr = io.EOF
+					req := <-client.send
+					// This can be plumbed through on either side, write or read,
+					// the important part is that we get it.
+					client.recvErr = expectedError
+					client.recv <- &pb.StateResponse{
+						Id: req.Id,
+					}
+				}()
+				_, err := c.Send(&pb.StateRequest{})
+				return err
+			},
+			expectedErr:       expectedError,
+			validateCancelled: true,
+		}, {
+			name: "writeOtherError",
+			caseFn: func(t *testing.T, c *StateChannel, client *fakeStateClient) error {
+				go func() {
+					client.sendErr = expectedError
+					<-client.send
+					// Shouldn't need to unblock any Recv calls.
+				}()
+				_, err := c.Send(&pb.StateRequest{})
+				return err
+			},
+			expectedErr:       expectedError,
+			validateCancelled: true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			client := &fakeStateClient{
+				recv: make(chan *pb.StateResponse),
+				send: make(chan *pb.StateRequest),
+			}
+			ctx, cancelFn := context.WithCancel(context.Background())
+			c := makeStateChannel(ctx, cancelFn, "id", client)
+			forceRecreateCalled := false
+			var forceRecreateError error
+			c.forceRecreate = func(_ string, err error) {
+				forceRecreateCalled = true
+				forceRecreateError = err
+			}
+
+			retErr := test.caseFn(t, c, client)
+
+			if got, want := retErr, test.expectedErr; !contains(got, want) {
+				t.Errorf("Unexpected error: got %v, want %v", got, want)
+			}
+
+			// Verify that new Sends return the same error on their reads after client.Recv is done.
+			go func() {
+				// Ensure that the client isn't helping us.
+				client.sendErr = nil
+				client.recvErr = nil
+				// Drain the next send, and ensure the response is unblocked.
+				req := <-client.send
+				client.recv <- &pb.StateResponse{Id: req.Id} // Ids need to match up to ensure routing can occur properly.
+			}()
+			if _, err := c.Send(&pb.StateRequest{}); !contains(err, test.expectedErr) {
+				t.Errorf("Unexpected error from Send: got %v, want %v", err, test.expectedErr)
+			}
+
+			if test.validateCancelled {
+				select {
+				case <-ctx.Done(): // Assert that the context must have been cancelled on read failures.
+					return
+				case <-time.After(time.Second * 5):
+					t.Fatal("context wasn't cancelled")
+				}
+				if !forceRecreateCalled {
```

Comment:
It looks like this check would never get reached because the select statement above either returns or calls a fatal error. Is the return supposed to be a break?

Reply:
Great spot! Thanks! The tests fortunately still pass. It's a bit redundant with the channel check, but I'd rather ensure that all places cancelFn is called, the forceRecreate is called.